### ISSUE TITLE
anndata optional dependency import system can behave poorly with autodoc_mock_imports

### ISSUE THREAD

**Issue by Intron7** [scverse/anndata#1134]
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [X] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

I ran into a problem with the new GPU dependencies and how they are loaded in anndata. For the rapids-singlecell docs I mock import all gpu packages, this includes cupy and cupyx. Because of the mock imports anndata thought it was in a GPU enviroment and tried to load the packages. The docs creation froze and didn't finish. Mock importing anndata didn't fix the issue it just resulted in a bug. What I had to do in the end is importing anndata at the start of the `conf.py` so that it doesn't know about cupy. If there is a more elegant solution than this please let me know. 

### Versions

anndata 0.10.0rc1

**Comment by ivirshup** [comment]
Hmm, I'm not 100% sure if I'd consider this a bug in anndata just yet, since this is definitely a strange use case.

Could you share how you did the mocking? And is this something that `cupy` recommends people do?

**Comment by Intron7** [comment]
I agree that it's not necessarily a bug. The behaviour is nevertheless unwanted and doesn't throw any meaningful errors. 
'''
autodoc_mock_imports = ["cudf", "cuml", "cugraph", "cupy", "cupyx"]
'''
This is how I mocked the gpu imports.

**Comment by flying-sheep** [comment]
Where is the code that checks if anndata is in a CUDA environment? I feel like it should be easy to change it so it doesn’t try to use it if it’s mocked.

**Comment by flying-sheep** [comment]
I think our current approach works well, including for Sphinx (it’ll correctly link `CupyArray` to cupy’s docs, as the `__name__` and `__module__` match)

```py
if find_spec("cupy") or TYPE_CHECKING:
    from cupy import ndarray as CupyArray
else:
    CupyArray = type("ndarray", (), dict(__module__="cupy"))
```

Please comment if we can improve anything.
